devmand/gateway/src/devmand/channels/cli/TimeoutTrackingCli.cpp

```diff
@@ -30,58 +30,82 @@ TimeoutTrackingCli::TimeoutTrackingCli(
-      executor(_executor),
-      timeoutInterval(_timeoutInterval) {
-  shutdown = false;
+    std::chrono::milliseconds _timeoutInterval) {
+  timeoutTrackingParameters =
+      shared_ptr<TimeoutTrackingParameters>(new TimeoutTrackingParameters{
+          _id, _cli, _timekeeper, _executor, _timeoutInterval, {false}});
 }
 
 TimeoutTrackingCli::~TimeoutTrackingCli() {
+  string id = timeoutTrackingParameters->id;
   MLOG(MDEBUG) << "[" << id << "] "
-               << "~TTCli";
-  shutdown = true;
-  executor = nullptr;
-  timekeeper = nullptr;
-  cli = nullptr;
+               << "~TTCli started";
+  timeoutTrackingParameters->shutdown = true;
+
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~TTCli nulling cli with refcount:"
+               << timeoutTrackingParameters->cli.use_count();
+  timeoutTrackingParameters->cli = nullptr;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~TTCli cli nulled";
+
+  while (timeoutTrackingParameters.use_count() >
+         1) { // TODO cancel currently running future
+    MLOG(MDEBUG) << "[" << timeoutTrackingParameters->id << "] "
+                 << "~TTCli sleeping";
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+  }
+
+  timeoutTrackingParameters = nullptr;
+
   MLOG(MDEBUG) << "[" << id << "] "
                << "~TTCli done";
 }
 
 Future<string> TimeoutTrackingCli::executeRead(const ReadCommand cmd) {
   return executeSomething(cmd, "TTCli.executeRead", [this, cmd]() {
-    return cli->executeRead(cmd);
+    return timeoutTrackingParameters->cli->executeRead(cmd);
   });
 }
 
 Future<string> TimeoutTrackingCli::executeWrite(const WriteCommand cmd) {
   return executeSomething(cmd, "TTCli.executeWrite", [this, cmd]() {
-    return cli->executeWrite(cmd);
+    return timeoutTrackingParameters->cli->executeWrite(cmd);
   });
 }
 
 Future<string> TimeoutTrackingCli::executeSomething(
     const Command& cmd,
     const string&& loggingPrefix,
     const function<Future<string>()>& innerFunc) {
-  MLOG(MDEBUG) << "[" << id << "] " << loggingPrefix << "('" << cmd
-               << "') called";
+  MLOG(MDEBUG) << "[" << timeoutTrackingParameters->id << "] (" << cmd.getIdx()
+               << ") " << loggingPrefix << "('" << cmd << "') called";
+  if (timeoutTrackingParameters->shutdown) {
+    return Future<string>(runtime_error("TTCli Shutting down"));
+  }
   Future<string> inner =
       innerFunc(); // we expect that this method does not block
-  if (shutdown)
-    throw runtime_error("TTCli Shutting down");
+  MLOG(MDEBUG) << "[" << timeoutTrackingParameters->id << "] (" << cmd.getIdx()
+               << ") "
+               << "Obtained future from underlying cli";
   return move(inner)
-      .via(executor.get())
+      .via(timeoutTrackingParameters->executor.get())
+      .thenValue(
+          [params = timeoutTrackingParameters, cmd](string result) -> string {
+            MLOG(MDEBUG) << "[" << params->id << "] (" << cmd.getIdx() << ") "
```

Comment:
we still need to log the command, not just the index. So we should log it via << cmd

and change the operator << implementation to provide e.g. "[11]show running-config"